File: algorithms/0074-search-a-2d-matrix/Solution.kt
// Search a 2D Matrix - flatten to 1D binary search; mid maps to (mid/cols, mid%cols).
// Time O(log(m*n)), Space O(1).

class Solution {
    fun searchMatrix(matrix: Array<IntArray>, target: Int): Boolean {
        val m = matrix.size
        val n = matrix[0].size
        var left = 0
        var right = m * n - 1
        while (left <= right) {
            val mid = left + (right - left) / 2
            val row = mid / n
            val col = mid % n
            val value = matrix[row][col]
            when {
                value == target -> return true
                value < target -> left = mid + 1
                else -> right = mid - 1
            }
        }
        return false
    }
}

File: algorithms/0074-search-a-2d-matrix/solution.py
# Search a 2D Matrix - flatten to 1D binary search; mid maps to (mid/cols, mid%cols).
# Time O(log(m*n)), Space O(1).

class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        m = len(matrix)
        n = len(matrix[0])
        left = 0
        right = m * n - 1
        while left <= right:
            mid = left + (right - left) // 2
            row = mid // n
            col = mid % n
            value = matrix[row][col]
            if value == target:
                return True
            if value < target:
                left = mid + 1
            else:
                right = mid - 1
        return False
